### PR TITLE
Tighten checks for "bins" elements in width_bucket(x, bins)

### PR DESCRIPTION
**Pull request by spershin**
## Description
Make width_bucket(x, bins) throw error if it finds a `null` or non-finite element in `bins`.

## Motivation and Context
Solves https://github.com/prestodb/presto/issues/24055

## Impact
Changes behavior of [width_bucket(x, bins)](https://prestodb.io/docs/current/functions/math.html#width_bucket-x-bound1-bound2-n-bigint) which previously was treating all `null` elements in `bins` as zero. Now the function will throw an error.

## Test Plan
Updated unit test to handle few cases with nulls.

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
```
== RELEASE NOTES ==
General Changes
* Fix behavior of :func:`width_bucket(x, bins) -> bigint` which previously treated all ``null`` elements in bins as ``0``. Now the function will throw an error if it finds a ``null`` or non-finite element in ``bins``..  :pr:`24103`
```

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java [hunk 1]
```diff
@@ -1335,7 +1342,7 @@ public void testCosineSimilarity()
 
         assertFunction("cosine_similarity(map(array ['a', 'b', 'c'], array [1.0E0, 2.0E0, -1.0E0]), map(array ['c', 'b'], array [1.0E0, 3.0E0]))",
                 DOUBLE,
-                (2 * 3 + (-1) * 1) / (Math.sqrt(1 + 4 + 1) * Math.sqrt(1 + 9)));
+                (2 * 3 + (-1)) / (Math.sqrt(1 + 4 + 1) * Math.sqrt(1 + 9)));
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:1345]: this looks like an unrelated change.

File: presto-docs/src/main/sphinx/functions/math.rst [hunk 2]
```diff
@@ -169,7 +169,8 @@ Mathematical Functions
 
     Returns the bin number of ``x`` according to the bins specified by the
     array ``bins``. The ``bins`` parameter must be an array of doubles and is
-    assumed to be in sorted ascending order.
+    assumed to be in sorted ascending order. The function will throw error if
+    encountered a ``null`` element in ``bins``.
```
Comment (rschlussel) [inline presto-docs/src/main/sphinx/functions/math.rst:173]:
```suggestion
    it encounters a ``null`` or non-finite element in ``bins``.
```

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java [hunk 3]
```diff
@@ -1574,11 +1574,17 @@ public static long widthBucket(@SqlType(StandardTypes.DOUBLE) double operand, @S
...
         while (lower < upper) {
+            if (bins.isNull(lower) || bins.isNull(upper - 1)) {
+                throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Bin values cannot be NULL");
+            }
             if (DOUBLE.getDouble(bins, lower) > DOUBLE.getDouble(bins, upper - 1)) {
                 throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Bin values are not sorted in ascending order");
             }
 
             index = (lower + upper) / 2;
+            if (bins.isNull(index)) {
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:1585]: the null checks showing up 2x makes me wonder if we should also be checking isFinite on lower/upper and checking sortedness for index (lower < index < upper).

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java (hunk 1 above)
Comment (spershin) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:1345]: Yes, I saw a warning in IDE that * 1 is not needed.
Decided it is small and harmless enough to add.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 3 above)
Comment (spershin) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java:1585]: We can do that.
Do we want it?
Then the scope of this change will grow past just nulls.

File: presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java (hunk 1 above)
Comment (rschlussel) [reply presto-main/src/test/java/com/facebook/presto/operator/scalar/TestMathFunctions.java:1349]: I think it's written that way to make the formula more clear. So you can understand where the results are coming from

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java (hunk 3 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: probably we do want it, but we can leave this change to just nulls.

Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java]: oh never mind, saw you updated already. Thanks!

**Review by rschlussel — DISMISSED** [review]
I'd undo the * 1 change for clarity, but everything else looks good

File: presto-docs/src/main/sphinx/functions/math.rst [hunk 4]
```diff
@@ -168,8 +168,11 @@ Mathematical Functions
 .. function:: width_bucket(x, bins) -> bigint
 
     Returns the bin number of ``x`` according to the bins specified by the
-    array ``bins``. The ``bins`` parameter must be an array of doubles and is
-    assumed to be in sorted ascending order.
+    array ``bins``. The ``bins`` parameter must be an array of doubles, should not
+    contain ``null`` or non-finite elements and assumed to be in sorted ascending order.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/functions/math.rst:172]:
```suggestion
    contain ``null`` or non-finite elements, and is assumed to be in sorted ascending order.
```

File: presto-docs/src/main/sphinx/functions/math.rst [hunk 5]
```diff
@@ -168,8 +168,11 @@ Mathematical Functions
...
+    The function will generally throw an error if it encounters a ``null`` or non-finite
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/functions/math.rst:173]:
```suggestion

    Note: The function returns an error if it encounters a ``null`` or non-finite
```
Suggest adding a line return to left-justify (within the width_bucket entry) a Note that describes the behavior of the function when it encounters a ``null``, and also what and why that error might not be returned.

File: presto-docs/src/main/sphinx/functions/math.rst [hunk 6]
```diff
@@ -168,8 +168,11 @@ Mathematical Functions
...
+    element in ``bins``. However, due to the binary search algorithm some such elements
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/functions/math.rst:174]:
```suggestion
    element in ``bins``, but due to the binary search algorithm some such elements
```

**Comment by steveburnett** [comment]
Because this PR changes user-facing behavior, I think we should consider a release note entry for this PR. Perhaps something like: 

```
== RELEASE NOTES ==
General Changes
* Fix behavior of :func:`width_bucket(x, bins) -> bigint` which previously treated all ``null`` elements in bins as ``0``. Now the function will throw an error.  :pr:`24103`
```

(I did a local doc build to verify that :func:`width_bucket(x, bins) -> bigint` generates a working link from a file in /release/ to this function in functions/math.rst.)

**Review by rschlussel — APPROVED** [review]

**Review by steveburnett — APPROVED** [review]
LGTM! (docs) 

Pull updated branch, new local docs build, looks good. Thank you for the quick revision!

**Comment by steveburnett** [comment]
Nit on the formatting of the release note entry - please add a row of three ` above and below the release note block. Like this: 

```
== RELEASE NOTES ==
General Changes
* Fix behavior of :func:`width_bucket(x, bins) -> bigint` which previously treated all ``null`` elements in bins as ``0``. Now the function will throw an error if it finds a ``null`` or non-finite element in ``bins``..  :pr:`24103`
```
